src/wp-includes/block-editor.php

```diff
@@ -297,3 +297,73 @@ function get_block_editor_settings( $editor_name, $custom_settings = array() ) {
 
 	return $editor_settings;
 }
+
+/**
+ * Preloads common data used with the block editor by specifying an array of
+ * REST API paths that will be preloaded for a given editor type.
+ *
+ * @since 5.8.0
+ *
+ * @global WP_Post $post Global post object.
+ *
+ * @param string $editor_name   The name of the editor (e.g. 'post-editor').
+ * @param array  $preload_paths List of paths to preload.
+ *
+ * @return void
+ */
+function block_editor_preload_api_requests( $editor_name, $preload_paths ) {
+	global $post;
+
+	/**
+	 * Filters the array of REST API paths that will be used to preloaded common data
+	 * to use with the block editor.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param string[] $preload_paths Array of paths to preload.
+	 */
+	$preload_paths = apply_filters( "block_editor_preload_paths_{$editor_name}", $preload_paths );
```

Comment:
Does this effectively mean you can no longer conditionally change what is preloaded based on the post type being edited? That's a substantial change.

Reply:
You can always use `get_post()` in the filter. In the Gutenberg plugin, we started simple, and there is only a filter that works with all editor types. We still need to figure out how to make it all simpler to use. The most pressing issue is that `$post` is not present on the widgets editor page so existing filters that have unguarded `$post` usage would error there.

Reply:
@gziolo That works until you start pulling the settings in via REST, and the API request doesn't have a post context :)

Reply:
The REST API endpoint is the very early stage of development. It might be a really good idea to provide more params to the API call so it could re-create the same behavior as you currently have when loading the edit post editor in WP-Admin.

Reply:
That'd be fantastic @gziolo :) Just want to make sure it's on the radar! 🚀

Reply:
Yeah, I'd echo this. We really shouldn't be encouraging developers trying to find the context out themselves. It can become a BC and maintenance nightmare since any change in the global context means that code will no longer work. We should really be passing a context object to this filter, not a `WP_Post` for instance, but a custom context class that has properties for things like a post.

Reply:
@TimothyBJacobs, @chrisvanpatten, would something like the simplified code below work here?

```php
<?php

class Block_Editor_Context {
    public $name = null;
    public $post = null;
    public function __get( $name ) {
        return null === $this->post ? null : $this->post->$name; 
    }
    public function __isset( $name ) {
        return isset( $this->post->$name );
    }
}
```

I see overloading methods used in other places so that seem like a viable approach here. The biggest benefit would be that we wouldn't have to deprecate existing filters but instead we pass the context instance in all places where `$post` was present.

Reply:
I wouldn't proxy it because if someone is expecting a `WP_Post` instance their site may fatal. I'd rather something like this where we define the possible types and have properties for the additional context values that may be relevant.

```php
class Block_Editor_Context {
    const POST = 'post-editor';
    const SITE = 'site-editor';
    const WIDGET = 'widget-editor';

    private $type;
    private $post;

    function get_type() { return $this->type }

    function get_post() { return $this->post }
}
```

Reply:
The type as presented here doesn't make sense to me. The line between post-editor and site-editor is very blurry in the frontend, they might show up at the same time, I'm not sure we can make that distinction on the server.

The post is not always present so if we do provide a context class like this, plugin authors should know that it's nullable.

Reply:
My understanding from the block editor settings stuff was that we did have different editor types. The line may be blurry, but isn't there a line? Particularly for the widgets and navigation?

If we don't want to have distinct types, should we be thinking in terms of "features" that the different editors have?

Reply:
> My understanding from the block editor settings stuff was that we did have different editor types. The line may be blurry, but isn't there a line?

For widgets, navigation yes, but not necessarily between post editor and site editor, you'd be able to navigate between these two seamlessly using the same settings in the template editor. There are also design proposals to split the site editor to multiple screens. I think it's too early to make assumptions about a "type" since these screens are still evolving.

Reply:
Yeah that makes sense. I do think though we need to be looking more at how editors get their settings. I should still be able to limit the blocks that show up when a user is editing a post, even if they are doing it thru the site editor.

If the screens do get split in the future, I think we can then add more types, no?

Reply:
> If the screens do get split in the future, I think we can then add more types, no?

I'm not sure it makes sense to have different settings between site editor and post editor even after splitting to multiple screens. To be honest, for me, it's very hard to reason about this. I know that for Core itself, we don't really need different settings but it's hard to know how it should scale to third-party plugins and in the future.

Reply:
I'm not sure I strongly care about distinguishing which editor is visible — I can understand how that might be ambiguous — but I do care about what specifically is being edited in terms of discrete pieces of site content (pages, posts, etc.) If that means that the context is "hey you're editing template part X" rather than "hey you're in the site editor" I think that's entirely reasonable; the context becomes about the edited object, which I think is the valuable bit.

I will say my _primary_ interest here is being able to restrict available blocks based on what is being edited, although I can see, for example, wanting to say "allow custom spacing when editing a template part, but not when editing a specific piece of content". But that's less concerning to me vs. solving for block restrictions based on the environment you're in/what you're editing.

Reply:
With https://github.com/WordPress/wordpress-develop/commit/0b6eb8daba605ca6c7e9b31200fe8b808f1b1465 I removed editor type specific filters for block editor configuration. It mostly aligns with what we have in the Gutenberg plugin at the moment. The planned follow-up is to replace the editor name string with a context object to bring back an optional reference to the actual post as discussed in this thread.

Reply:
@chrisvanpatten that's a very good way to think about it and it's clear and well defined that I can agree with.

The issue with that is that sometimes the "edited piece of content" changes dynamically inside the same editor instance.

A template with a Query block can edited multiple posts and single template at the same time which means a logic like you suggest would only work if these editor settings were specific to a given block container (InnerBlocks). Right now, these settings are unique in a single block editor instance and provided using php (inline script) so it's not something we can support at the moment unless we perform a big JS refactoring. (not saying it's not doable but I wanted to get expectations right :) )

Reply:
@youknowriad I totally get that. I do think longer term, there is value in InnerBlocks or other block containers being able to fetch their own settings at run-time — I've certainly always advocated for more control over InnerBlock containers 😅  — but I think it's also reasonable as a first step that the editor settings default to whatever object is being edited at launch and there is an inherent drift as you move through different screens and experiences.

Thinking on a longer horizon, I wonder if the editor settings are turning to a bit of a catch-all… maybe editor settings should be split into more logical groups, some of which are persistent across _all_ editor experiences, and others which are contextual? That might make the whole thing a little easier to reason about and also potentially make it easier to load contextual stuff up front.

I'd also emphasize that for the enterprise use-cases I've been involved with (at Meredith and now at the NBA), we've generally been happy with object _type_ level contextual fidelity. In other words, we're rarely (never?) saying "we need to configure settings for this individual specific post", but "we need to configure settings for post type X". Because we can't necessarily predict which specific _objects_ might be edited in a given editor session, maybe it's more reasonable to pre-load settings for all object _types_ up front instead?

Reply:
I included `WP_Block_Editor_Context` class that only allows setting the `$post` field for now. If we pass it to all places, we will be able to grow the class as we add new functionalities so we can decide based on individual cases. This way we don't have to predict what names of the editors we should use. It also gives the flexibility to the REST API endpoint to set the context in an expected state. 

I'm planning to add tests today and integrate everything with the block editor settings functions. I feel like it satisfies all the feedback shared. If there are no further concerns I'd like to commit changes on Monday to have it ready for the WP 5.8 feature freeze.